### PR TITLE
Feature/1667/add path to node context menu

### PR DESCRIPTION
**Pull request by Shivan-997**
# add path to node context menu

closes: #1667 

## Description

Path of node is added to the node context menu and is fully visible when hovering on the path.

## Screenshots or gifs
Screenshot 2021-04-19 at 11 52 01

### CODE REVIEW

File: visualization/app/codeCharta/ui/nodeContextMenu/nodeContextMenu.component.ts [hunk 1]
```diff
@@ -67,6 +71,9 @@ export class NodeContextMenuController
 
 	onShowNodeContextMenu(path: string, nodeType: string, mouseX: number, mouseY: number) {
 		this._viewModel.codeMapNode = getCodeMapNodeFromPath(path, nodeType, this.codeMapPreRenderService.getRenderMap())
+		this._viewModel.nodePath = path
+		this._viewModel.lastPartOfNodePath =
+			path.lastIndexOf("/") === 0 ? path.slice(path.lastIndexOf("/")) : `...${path.slice(path.lastIndexOf("/"))}`
```
Comment (NearW) [inline visualization/app/codeCharta/ui/nodeContextMenu/nodeContextMenu.component.ts:76]:
```suggestion
		this._viewModel.lastPartOfNodePath = `${path.lastIndexOf("/") === 0 ? "" : "..."}${path.slice(path.lastIndexOf("/"))}`
```

**Comment by NearW** [comment]
@Shivan-997 LGTM!
Can we write a test for both cases to complete the story?

Test-Cases:
"/root/folder/file.jpg" -> ".../file.jpg"
"/root" -> "/root"

File: visualization/app/codeCharta/ui/nodeContextMenu/nodeContextMenu.component.spec.ts [hunk 2]
```diff
@@ -165,6 +165,34 @@ describe("nodeContextMenuController", () => {
 			expect(document.getElementById).toHaveBeenCalledWith("codeMap")
 			expect(elementMock.addEventListener).toHaveBeenCalledWith("wheel", expect.anything(), expect.anything())
 		})
+
+		it("should not shorten the path if it has no sub paths", () => {
+			document.body.addEventListener = jest.fn()
+
+			const elementMock = { addEventListener: jest.fn() }
+			// @ts-ignore
+			jest.spyOn(document, "getElementById").mockImplementation(() => elementMock)
```
Comment (NearW) [inline visualization/app/codeCharta/ui/nodeContextMenu/nodeContextMenu.component.spec.ts:174]: move these to beforeEach and remove it from each test

Comment (Shivan-997) [reply visualization/app/codeCharta/ui/nodeContextMenu/nodeContextMenu.component.spec.ts:174]: expect(elementMock.addEventListener).toHaveBeenCalledWith("wheel", expect.anything(), expect.anything()) 
complains that element mock is not defined if i move it to befor each

Comment (NearW) [reply visualization/app/codeCharta/ui/nodeContextMenu/nodeContextMenu.component.spec.ts:174]: Just move it to the top then:

```ts
describe("onShowNodeContextMenu", () => {
  let elementMock
  beforeEach(() => {
    nodeContextMenuController.setPosition = jest.fn()
    nodeContextMenuController.calculatePosition = jest.fn().mockReturnValue({ x: 1, y: 2 })

    document.body.addEventListener = jest.fn()
    elementMock = { addEventListener: jest.fn() }
    // @ts-ignore
    jest.spyOn(document, "getElementById").mockImplementation(() => elementMock)
  })
  it("should set the correct building after some timeout", () => {
    nodeContextMenuController.onShowNodeContextMenu("/root", NodeType.FOLDER, 42, 24)
    
    expect(nodeContextMenuController["_viewModel"].codeMapNode).toEqual(TEST_DELTA_MAP_A.map)
    expect(nodeContextMenuController["_viewModel"].showNodeContextMenu).toBe(true)
    expect(nodeContextMenuController.calculatePosition).toHaveBeenCalledWith(42, 24)
    expect(nodeContextMenuController.setPosition).toHaveBeenCalledTimes(1)
    expect(nodeContextMenuController.setPosition).toBeCalledWith(1, 2)
    expect(document.body.addEventListener).toHaveBeenNthCalledWith(1, "click", expect.anything(), expect.anything())
    expect(document.body.addEventListener).toHaveBeenNthCalledWith(2, "mousedown", expect.anything(), expect.anything())
    expect(document.getElementById).toHaveBeenCalledWith("codeMap")
    expect(elementMock.addEventListener).toHaveBeenCalledWith("wheel", expect.anything(), expect.anything())
  })
})
```

**Review by NearW — APPROVED** [review]

**Pull request merged**